src/marker.js

```diff
@@ -126,7 +126,26 @@ function applyMarkersColor(path, {markerStart, markerMid, markerEnd, stroke}, st
     };
   }
 
-  if (markerStart) path.attr("marker-start", applyMarker(markerStart));
-  if (markerMid) path.attr("marker-mid", applyMarker(markerMid));
-  if (markerEnd) path.attr("marker-end", applyMarker(markerEnd));
+  if (!(markerStart || markerMid || markerEnd)) return;
+
+  const start = markerStart && applyMarker(markerStart);
+  const mid = markerMid && applyMarker(markerMid);
+  const end = markerEnd && applyMarker(markerEnd);
+  if (Z) {
+    for (const g of group(
+      path.filter((i) => i.length >= 2),
```

Comment:
I wonder if segments of length 1 are not a bug already—independently of the grouping?

Reply:
Could be. The only case I noticed was at the end of a line, which seems unavoidable. Are there others?

Reply:
No, I think that's the only case. It makes "broken arrows" (if we apply a marker-end: arrow), which is why we have to filter it out here. I'll investigate.

Reply:
Here’s a patch which “fixes” it. But I think it’s valid for the last segment to have a single index if it has different aesthetics. It might be visible in the case where the stroke-linecap is round or square.

```patch
diff --git a/src/style.js b/src/style.js
index a1af100f..420d7bd3 100644
--- a/src/style.js
+++ b/src/style.js
@@ -261,6 +261,7 @@ export function* groupIndex(I, position, mark, channels) {
   for (const G of Z ? groupZ(I, Z, z) : [I]) {
     let Ag; // the A-values (aesthetics) of the current group, if any
     let Gg; // the current group index (a subset of G, and I), if any
+    let singleton = false; // does this series only have one group?
     out: for (const i of G) {
       // If any channel has an undefined value for this index, skip it.
       for (const c of C) {
@@ -273,8 +274,8 @@ export function* groupIndex(I, position, mark, channels) {
       // Otherwise, if this is a new group, record the aesthetics for this
       // group. Yield the current group and start a new one.
       if (Ag === undefined) {
-        if (Gg) yield Gg;
         (Ag = A.map((c) => keyof(c[i]))), (Gg = [i]);
+        singleton = true;
         continue;
       }
 
@@ -287,13 +288,15 @@ export function* groupIndex(I, position, mark, channels) {
         if (k !== Ag[j]) {
           yield Gg;
           (Ag = A.map((c) => keyof(c[i]))), (Gg = [i]);
+          singleton = false;
           continue out;
         }
       }
     }
 
-    // Yield the current group, if any.
-    if (Gg) yield Gg;
+    // Yield the current group, if any. For non-singleton series, the last group
+    // might contain only a single index; this is ignored.
+    if (Gg?.length > 1 || singleton) yield Gg;
   }
 }
 
```

Reply:
When the last segment has only one point the linecap has the default ("horizontal") orientation, which makes it feel broken (here, shown with transparency so we can see better):

<img width="163" alt="Capture d’écran 2024-06-18 à 08 39 29" src="https://github.com/observablehq/plot/assets/7001/9eb3fc03-5a9f-473d-90b9-5429a74193c2">

With your patch it shows as I expect:
<img width="204" alt="Capture d’écran 2024-06-18 à 08 44 06" src="https://github.com/observablehq/plot/assets/7001/14767105-5ad9-4ab0-9f78-6133913322a9">

Reply:
I think I want to apply this patch! Not strictly necessary for this PR, but it seems reasonable to do it, and it will clean up the awkward tests on `I.length > 1`

Reply:
I’m not convinced we want the patch… Let’s land this without it and revisit if needed.